xbmc/Application.cpp

```diff
@@ -1978,7 +1979,34 @@ bool CApplication::OnAction(const CAction &action)
     return true;
   }
 
-  if ((action.GetID() == ACTION_INCREASE_RATING || action.GetID() == ACTION_DECREASE_RATING) && m_appPlayer.IsPlayingAudio())
+  if ((action.GetID() == ACTION_SET_RATING) && m_appPlayer.IsPlayingAudio())
+  {
+    const CMusicInfoTag *tag = g_infoManager.GetCurrentSongTag();
+    if (tag)
+    {
+      //Update with any rating changes to playing song already made via GUI e.g. on song info dialog
+      *m_itemCurrentFile->GetMusicInfoTag() = *tag;
+      
+      int userrating = MUSIC_UTILS::ShowSelectRatingDialog(m_itemCurrentFile->GetMusicInfoTag()->GetUserrating());
+      if (userrating < 0) // Nothing selected, so user rating unchanged
+        return true;
+      userrating = std::min(userrating, 10);
+      if (userrating != m_itemCurrentFile->GetMusicInfoTag()->GetUserrating())
+      {
+        m_itemCurrentFile->GetMusicInfoTag()->SetUserrating(userrating);
+        // Send a message to all windows to tell them to update the fileitem in 
+        // particular to ensure that g_infoManager.m_currentFile is kept a mirror of m_itemCurrentFile
+        CGUIMessage msg(GUI_MSG_NOTIFY_ALL, 0, 0, GUI_MSG_UPDATE_ITEM, 0, m_itemCurrentFile);
```

Comment:
all gui items should update from GUIInfoManager. I would update GUIInfomanager:
CGUIInfoManager::SetCurrentItem
and have it send the gui message when update is done.